controllers/istiorevision/istiorevision_controller.go

```diff
@@ -372,6 +400,109 @@ func (r *IstioRevisionReconciler) determineReadyCondition(ctx context.Context,
 	}, nil
 }
 
+func (r *IstioRevisionReconciler) determineInUseCondition(ctx context.Context, rev *v1alpha1.IstioRevision) (v1alpha1.IstioRevisionCondition, error) {
+	isReferenced, err := r.isRevisionReferencedByWorkloads(ctx, rev)
+	if err != nil {
+		return v1alpha1.IstioRevisionCondition{}, err
+	}
+
+	if isReferenced {
+		return v1alpha1.IstioRevisionCondition{
+			Type:    v1alpha1.IstioRevisionConditionTypeInUse,
+			Status:  metav1.ConditionTrue,
+			Reason:  v1alpha1.IstioRevisionConditionReasonReferencedByWorkloads,
+			Message: "The IstioRevision is referenced by at least one pod or namespace",
+		}, nil
+	}
+	return v1alpha1.IstioRevisionCondition{
+		Type:    v1alpha1.IstioRevisionConditionTypeInUse,
+		Status:  metav1.ConditionFalse,
+		Reason:  v1alpha1.IstioRevisionConditionReasonNotReferenced,
+		Message: "The IstioRevision is not referenced by any pod or namespace",
+	}, nil
+}
+
+func (r *IstioRevisionReconciler) isRevisionReferencedByWorkloads(ctx context.Context, rev *v1alpha1.IstioRevision) (bool, error) {
+	logger := log.FromContext(ctx)
+	nsList := corev1.NamespaceList{}
+	nsMap := map[string]corev1.Namespace{}
+	if err := r.Client.List(ctx, &nsList); err != nil { // TODO: can we optimize this by specifying a label selector
+		return false, err
+	}
+	for _, ns := range nsList.Items {
+		if namespaceReferencesRevision(ns, rev) {
+			logger.Info("Revision is referenced by Namespace", "Namespace", ns.Name)
+			return true, nil
+		}
+		ns := ns
+		nsMap[ns.Name] = ns
+	}
+
+	podList := corev1.PodList{}
+	if err := r.Client.List(ctx, &podList); err != nil { // TODO: can we optimize this by specifying a label selector
+		return false, err
+	}
+	for _, pod := range podList.Items {
+		if ns, found := nsMap[pod.Namespace]; found && podReferencesRevision(pod, ns, rev) {
+			logger.Info("Revision is referenced by Pod", "Pod", client.ObjectKeyFromObject(&pod))
+			return true, nil
+		}
+	}
+
+	values := rev.Spec.GetValues()
+	if enableNamespacesByDefault, _, err := values.GetBool("sidecarInjectorWebhook.enableNamespacesByDefault"); err != nil {
+		return false, err
+	} else if enableNamespacesByDefault && rev.Name == "default" {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func namespaceReferencesRevision(ns corev1.Namespace, rev *v1alpha1.IstioRevision) bool {
+	return rev.Name == getReferencedRevisionFromNamespace(ns.Labels)
+}
+
+func podReferencesRevision(pod corev1.Pod, ns corev1.Namespace, rev *v1alpha1.IstioRevision) bool {
+	return rev.Name == getReferencedRevisionFromPod(pod.GetLabels(), pod.GetAnnotations(), ns.GetLabels())
+}
+
+func getReferencedRevisionFromNamespace(labels map[string]string) string {
+	if labels["istio-injection"] == "enabled" {
+		return "default"
+	}
+	revision := labels["istio.io/rev"]
+	if revision != "" {
+		return revision
+	}
+	// TODO: if .Values.sidecarInjectorWebhook.enableNamespacesByDefault is true, then all namespaces except system namespaces should use the "default" revision
```

Comment:
Should we create tickets for these TODO's so we don't forget about them?

Reply:
I created https://issues.redhat.com/browse/OSSM-5572 to address these and any future TODOs.